bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudAccountBridgeHandler.java

```diff
@@ -88,85 +101,281 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                    Thread.sleep(200);
+                } catch (InterruptedException e) {
+                    Thread.interrupted();
+                }
+            } while (!success && retryCount < 3);
+            throw new RuntimeException("Invocation failed finally.", lastException);
+        }
+    }
+
+    /**
+     * @throws InterruptedException
+     * @throws IOException
+     *
+     */
+    private boolean handle2FAAuthentication() throws IOException, InterruptedException {
+        logger.debug("Starting iCloud 2-FA authentication  AuthState={}, Thing={})...", authState,
+                getThing().getUID().getAsString());
+        if (authState != AuthState.WAIT_FOR_CODE) {
+            throw new IllegalStateException("2-FA authentication not initialized.");
+        }
+        ICloudAccountThingConfiguration config = getConfigAs(ICloudAccountThingConfiguration.class);
+        String lastTriedCode = storage.get(AUTH_CODE_KEY);
+        boolean success = false;
+        if (config.code == null || config.code.isBlank() || config.code.equals(lastTriedCode)) {
+            // Still waiting for user to update config.
+            logger.warn("ICloud authentication requires 2-FA code. Please provide code configuration for thing '{}'.",
+                    getThing().getUID().getAsString());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Please provide 2-FA code in thing configuration.");
+            return false;
+        } else {
+            // 2-FA-Code was requested in previous call of this method.
+            // User has provided code in config.
+            logger.debug("Code is given in thing configuration '{}'. Trying to validate code...",
+                    getThing().getUID().getAsString());
+            storage.put(AUTH_CODE_KEY, lastTriedCode);
+            success = this.iCloudService.validate2faCode(config.code);
+            if (!success) {
+                authState = AuthState.INITIAL;
+                logger.warn("ICloud token invalid.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid 2-FA-code.");
+                return false;
+            }
+            org.openhab.core.config.core.Configuration config2 = editConfiguration();
+            config2.put("code", "");
+            updateConfiguration(config2);
+
+            logger.debug("Code is valid.");
+        }
+        authState = AuthState.AUTHENTICATED;
+        updateStatus(ThingStatus.ONLINE);
+        logger.debug("iCloud bridge handler '{}' authenticated with 2-FA code.", getThing().getUID().getAsString());
+        return success;
     }
 
     @Override
     public void handleRemoval() {
+
         super.handleRemoval();
     }
 
     @Override
     public void dispose() {
-        if (refreshJob != null) {
-            refreshJob.cancel(true);
+        if (this.refreshJob != null) {
+            this.refreshJob.cancel(true);
```

Comment:
Why this change? Probably there was already a null annotation warning, which could be fixed like this:
```suggestion
        ScheduledFuture<?> refreshJob = this.refreshJob;
        if (refreshJob != null) {
            refreshJob.cancel(true);
            this.refreshJob = null;
```